### DISCUSSION TITLE
How to verify the `tryExitOffer` execution of a `bad invitation`?

### DISCUSSION THREAD

**Discussion by Jorge-Lopes** [Agoric/agoric-sdk#10374] [Q&A] [answered]
# Question
The purpose of this discussion is to find an approach to verify that a transaction that executes a `tryExitOffer` to reclaim the payment of a `bad invitation` was successfully executed.

# Context
In the context of a3p acceptance tests, there is one test case that validates that an [offer](https://github.com/Agoric/agoric-3-proposals/blob/main/proposals/74%3Aupgrade-15/prepare.sh) built with an invalid method during upgrade-15 can be exited and the payment redeemed.

By comparing the user's balance, we can verify that this is the case. Although, the current implementation relies on the method `waitForBlock`, which is not ideal, as discussed [here](https://github.com/Agoric/agoric-sdk/pull/10123/files#r1772319519).

At [smatWallet.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/smart-wallet/src/smartWallet.js ), the [executeOffer](https://github.com/Agoric/agoric-sdk/blob/master/packages/smart-wallet/src/smartWallet.js#L890-L990) method will call `withdrawGive` and add the payment to `liveOfferPayments` , but it will fail before adding the offer to `liveOffers` and `liveOfferSeats`.

This means that when [tryExitOffer](https://github.com/Agoric/agoric-sdk/blob/master/packages/smart-wallet/src/smartWallet.js#L998-L1016) method is called, the payment will be reclaimed but then, it will throw with a message similar to 
`Error: key bad-invitation-15 not found in collection live offer seats`.

# Problem

In the effort to find a replacement for waitForBlock, I started by checking if the user's wallet, `liveOffers,` would contain the bad-invitation-15 before the tryExitOffer call is executed. As explained above, that is not the case.

During today's office hours, @dckc suggested that the status of the `tryExitOffer` call could be visible in the Vstorage node `published.wallet.{addr}`.

This does not seem to be the case either, as we can see from this [test](https://gist.github.com/Jorge-Lopes/a383be043aee25375789d4f0eeeae2eb). Which I assume it is because the tryExitOffer method throws because the offerId can't be found at liveOfferSeats.

# How to reproduce:

1. Checkout to branch `jorge/acceptance-replace-waitForBlock`.
2. Copy this [test](https://gist.github.com/Jorge-Lopes/a383be043aee25375789d4f0eeeae2eb) into `a3p-integration/proposals/z:acceptance`
3. Open terminal 1: 
```
> cd a3p-integration
> yarn build
> docker buildx bake --load test-acceptance
> docker run -p 26657:26657 -p 9090:9090 -p 1317:1317 -it --rm --name a3p -w /usr/src/a3p-integration/proposals/z:acceptance  --entrypoint bash --mount type=bind,src=.,dst=/usr/src/a3p-integration ghcr.io/agoric/agoric-3-proposals:test-acceptance
> source /usr/src/upgrade-test-scripts/env_setup.sh
```
4. Open terminal 2: 
```
> cd a3p-integration
> docker exec -w /usr/src/upgrade-test-scripts -it a3p bash start_agd.sh
```
5. Terminal 1: 
```
>  yarn ava {name_given_at_step_2}.js
```

# Logs

Chain logs:
```sh
2024-10-30T18:57:30.655Z SwingSet: vat: v43: walletFactory.fromBridge: { blockHeight: 1360, blockTime: 1730314649, owner: 'agoric1ee9hr0jyrxhy999y755mp862ljgycmwyp4pl7q', spendAction: '{"body":"#{\\"method\\":\\"tryExitOffer\\",\\"offerId\\":\\"bad-invitation-15\\"}","slots":[]}', type: 'WALLET_SPEND_ACTION' }
2024-10-30T18:57:30.656Z SwingSet: vat: v43: walletFactory: { wallet: Object [Alleged: SmartWallet self] {}, actionCapData: { body: '#{"method":"tryExitOffer","offerId":"bad-invitation-15"}', slots: [] } }
2024-10-30T18:57:30.701Z SwingSet: vat: v43: wallet agoric1ee9hr0jyrxhy999y755mp862ljgycmwyp4pl7q reclaimed [ { brand: Object [Alleged: IST brand] {}, value: 15_000n } ] from bad-invitation-15
2024-10-30T18:57:30.702Z SwingSet: ls: v43: Logging sent error stack (Error#1)
2024-10-30T18:57:30.702Z SwingSet: ls: v43: Error#1: key bad-invitation-15 not found in collection live offer seats
2024-10-30T18:57:30.702Z SwingSet: ls: v43: Error: key (a string) not found in collection "live offer seats"
```

Test logs:

```sh
LOG wallet node before:  {
        status: {
          id: 'pushPriceRound-1730282371904',
          invitationSpec: {
            invitationArgs: Array [ … ],
            invitationMakerName: 'PushPrice',
            previousOffer: 'stATOM.n-upgrade.0',
            source: 'continuing',
          },
          numWantsSatisfied: 1,
          payouts: {},
          proposal: {},
          result: undefined,
        },
        updated: 'offerStatus',
      }
```

```sh
LOG wallet node after:  {
        status: {
          id: 'pushPriceRound-1730282371904',
          invitationSpec: {
            invitationArgs: Array [ … ],
            invitationMakerName: 'PushPrice',
            previousOffer: 'stATOM.n-upgrade.0',
            source: 'continuing',
          },
          numWantsSatisfied: 1,
          payouts: {},
          proposal: {},
          result: undefined,
        },
        updated: 'offerStatus',
      }
```

```sh
broadcastBridgeAction { method: 'tryExitOffer', offerId: 'bad-invitation-15' }
Executing  [
  '--node=http://0.0.0.0:26657',
  '--chain-id=agoriclocal',
  '--keyring-backend=test',
  '--from=agoric1ee9hr0jyrxhy999y755mp862ljgycmwyp4pl7q',
  'tx',
  'swingset',
  'wallet-action',
  '--allow-spend',
  '{"body":"#{\\"method\\":\\"tryExitOffer\\",\\"offerId\\":\\"bad-invitation-15\\"}","slots":[]}',
  '--output',
  'json',
  '--yes'
]
2024-10-30T19:29:29.474074754Z tx not in block 1324 retrying...
```

**Comment by dckc** [comment, marked as answer]
That's unfortunate.

I suppose the test we used was to check the account balance.

https://github.com/Agoric/agoric-3-proposals/blob/c9644d6cdc28af8cc40c836ea17662935a948c9f/proposals/75%3Aupgrade-16/exit-reclaim.test.js#L23-L35

could you use a "retry until the balance changes" approach?

**Reply by Jorge-Lopes** [reply]
Thank you very much for the suggestion.
The implementation of this approach can be seen [here](https://github.com/Agoric/agoric-sdk/blob/jorge/acceptance-replace-waitForBlock/a3p-integration/proposals/z%3Aacceptance/wallet.test.js#L54-L76) and it worked as intended
